app/coffee/AuthHandler.coffee

```diff
@@ -0,0 +1,93 @@
+_ = require("lodash")
+async = require("async")
+bcrypt = require("bcrypt")
+crypto = require("crypto")
+logger = require("logger-sharelatex")
+
+{ db } = require("./MongoHandler")
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000    # 1 hour
+TOKEN_SEARCH_PREFIX_LENGTH = 16  # 8 bytes
+
+module.exports = AuthHandler =
+	login: (email, password, callback) ->
+		db.users.findOne {email: email.toLowerCase()}, (err, user) ->
+			if err?
+				return callback(err)
+			if !user
+				return callback(new Errors.AuthenticationError("user not found"))
+			if !user.hashedPassword
+				return callback(new Errors.AuthenticationError("user does not have a password"))
+			bcrypt.compare password, user.hashedPassword, (err, match) ->
+				if err?
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
+				callback(null, user._id.toString())
+
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		async.auto {
+			tokenDoc: (cb) =>
+				tokenPrefix = token.slice(0, TOKEN_SEARCH_PREFIX_LENGTH)
```

Comment:
I agree, but I think this validation goes at the controller level. I had this as a nice-to-have next PR. I now realize that maybe I should add it right away. I was thinking of introducing a validation library instead of doing it ad hoc.

At this level, we are protected from arbitrary regexps by `_.escapeRegExp()` in the next line.